ocaml/typing/typecore.ml

```diff
@@ -2413,27 +2464,24 @@ and type_pat_aux
   | Ppat_or(sp1, sp2) ->
       begin match mode with
       | Normal ->
-          let initial_pattern_variables = !pattern_variables in
-          let initial_module_variables = !module_variables in
+          let tps1 = tps in
+          let tps2 = copy_type_pat_state tps in
```

Comment:
Reusing the original state for `tps1` seems a bit confusing. I'd have expected us to make two copies here. It makes it easier to see that all the changes to the original state come from the merging process below.

Reply:
I changed it to make two copies. (The only reason I chose the old approach was that it seemed closer to the old impl, but I agree this seems conceptually more straightforward.)

Reply:
I'm clicking "resolve" because I took your suggestion, but feel free to reopen.